### PR TITLE
fix: Pass proxies config when using ClientSecretCredential in AzureDataLakeStorageV2Hook

### PR DESCRIPTION
**Pull request by dabla**
This pull request contains a fix when AzureDataLakeStorageV2Hook is used behing a corporate proxy, as the proxies config wasn't passed to the ClientSecretCredential in the AzureDataLakeStorageV2Hook the issue was that the bearer token could not be fetched.  Also added a DataToADLSOperator which uses the AzureDataLakeStorageV2Hook which allows uploading data (like from an XCOM of a previous task for example as we had the case) to a remote file without the need to create a local file first.  Both the fix and the new operator has unit test and docs where added for the DataToADLSOperator.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/providers/microsoft/azure/transfers/local_to_adls.py [hunk 1]
```diff
@@ -96,3 +96,53 @@ def execute(self, context: Context) -> None:
...
+    """
+    Upload data to Azure Data Lake.
+
+    .. seealso::
+        For more information on how to use this operator, take a look at the guide:
+        :ref:`howto/operator:DataToADLSOperator`
+
+    :param file_system_name: Name of the file system or instance of FileSystemProperties.
+    :param file_name: Name of the file which needs to be created in the file system.
+    :param data: Number of threads to use. If None, uses the number of cores.
```
Comment (Lee-W) [inline airflow/providers/microsoft/azure/transfers/local_to_adls.py]: The docstring here seems to be wrong

File: airflow/providers/microsoft/azure/transfers/local_to_adls.py [hunk 2]
```diff
@@ -96,3 +96,53 @@ def execute(self, context: Context) -> None:
...
+    :param length: Number of threads to use. If None, uses the number of cores.
```
Comment (Lee-W) [inline airflow/providers/microsoft/azure/transfers/local_to_adls.py]: Could we please check this one as well?

**Comment by dabla** [comment]
@Lee-W Thanks for pointing that out, I indeed did forget to update the docstrings, this is ok now.

**Review by Lee-W — APPROVED** [review]

**Review by eladkal — APPROVED** [review]

**Review by eladkal — CHANGES_REQUESTED** [review]
> Also added a DataToADLSOperator which uses the AzureDataLakeStorageV2Hook which allows uploading data (like from an XCOM of a previous task for example as we had the case) to a remote file without the need to create a local file first.

I am not sure if this is a valid use case (I think we had this discussion before on another operator).
What exactly are we saving by having DataToADLSOperator? xcom data is small so it can't be performance(?)

Lets please remove this part from the PR as it's not related to the bug fix
You can start a separated PR with adding the operator.

**Comment by dabla** [comment]
> > Also added a DataToADLSOperator which uses the AzureDataLakeStorageV2Hook which allows uploading data (like from an XCOM of a previous task for example as we had the case) to a remote file without the need to create a local file first.
> 
> I am not sure if this is a valid use case (I think we had this discussion before on another operator). What exactly are we saving by having DataToADLSOperator? xcom data is small so it can't be performance(?)
> 
> Lets please remove this part from the PR as it's not related to the bug fix You can start a separated PR with adding the operator.

Hello@eladkal ok I'll remove it but know that XCom's don't always have to be small.  Like in our case we have a custom XCom backend which stores large sets of data (like paged results of a REST API) on a PersistentVolumeClain, which means only the reference to the file is being stored in the Airflow database and the data itself resides on the filesystem, which is distributed accross the different workers.  That way we have operators which retrieves the data and store those results as XCom's, and the consecutive operator, here the DataToADLSOperator takes the result of that XCom and stores it on Fabric.  We could of course first write a python task which takes the XCom and writes it down as file before passing it to the Azure operator, but that would mean we need to invoke an additonal worker just to do that, while with this operator you can skip that additional step which also means less code to maintain and a more readable and concise DAG.

**Comment by eladkal** [comment]
>  I'll create a new PR for this DataToADLSOperator operator

Please do. We can discuss there

**Review by eladkal — APPROVED** [review]

**Pull request merged**